core-tests/src/test/java/org/mule/runtime/core/message/DefaultMuleMessageBuilderTestCase.java

```diff
@@ -0,0 +1,335 @@
+                (TEST_ATTR).build();
+
+        assertThat(message.getPayload(), is(TEST_PAYLOAD));
+        assertThat(message.getDataType(), is(STRING));
+        assertThat(message.getAttributes(), is(TEST_ATTR));
+    }
+
+    @Test
+    public void createNewMessageCollectionViaMessageInterface()
+    {
+        List<String> htmlStringList = new ArrayList<>();
+        htmlStringList.add("HTML1");
+        htmlStringList.add("HTML2");
+        htmlStringList.add("HTML3");
+
+        MuleMessage<List<String>, Apple> message;
+        message = MuleMessage.builder().collectionPayload(htmlStringList, String.class)
+                                       .itemMediaType(HTML)
+                                       .attributes(TEST_ATTR)
+                                       .build();
+
+        assertThat(message.getPayload(), is(htmlStringList));
+        assertThat(message.getDataType().getType(), equalTo(ArrayList.class));
+        assertThat(message.getDataType().getMediaType(), is(ANY));
+        assertThat(message.getDataType(), instanceOf(DefaultCollectionDataType.class));
+        assertThat(((DefaultCollectionDataType) message.getDataType()).getItemDataType().getMediaType(), equalTo(HTML));
+    }
+
+    @Test
+    public void createNewMessageCollectionViaMessageInterfaceCopy()
+    {
+        List<String> htmlStringList = new ArrayList<>();
+        htmlStringList.add("HTML1");
+        htmlStringList.add("HTML2");
+        htmlStringList.add("HTML3");
+
+        MuleMessage<List<String>, Apple> message;
+        message = MuleMessage.builder().collectionPayload(htmlStringList, String.class)
+                .itemMediaType(HTML)
+                .attributes(TEST_ATTR)
+                .build();
+
+        MuleMessage<List<String>, Apple> copy = MuleMessage.builder(message).build();
+
+        assertThat(copy.getPayload(), is(htmlStringList));
+        assertThat(copy.getDataType().getType(), equalTo(ArrayList.class));
+        assertThat(copy.getDataType().getMediaType(), is(ANY));
+        assertThat(copy.getDataType(), instanceOf(DefaultCollectionDataType.class));
+        assertThat(((DefaultCollectionDataType) copy.getDataType()).getItemDataType().getMediaType(), equalTo(HTML));
+    }
+
+    @Test
+    public void createMessageViaMessageInterfaceFromCopy()
+    {
+        MuleMessage<Boolean, Banana> messageCopy = MuleMessage.builder(createTestMessage()).payload(true)
+                .attributes(TEST_ATTR_2).build();
+
+        assertThat(messageCopy.getPayload(), is(true));
+        assertThat(messageCopy.getDataType(), is(BOOLEAN));
+        assertThat(messageCopy.getAttributes(), is(TEST_ATTR_2));
+    }
+
+    @Test
+    public void messageAttributes()
+    {
+        assertTestMessage(createTestMessage());
+    }
+
+    @Test
+    public void messageAttributesCopy()
+    {
+        assertTestMessage(new DefaultMuleMessageBuilder(createTestMessage()).build());
```

Comment:
would we prefer this method for getting this builder, instead of MuleMessage.builder() and then casting?